pin/gc/triset.go

```diff
@@ -0,0 +1,160 @@
+package gc
+
+import (
+	"context"
+
+	dag "github.com/ipfs/go-ipfs/merkledag"
+
+	cid "gx/ipfs/QmTprEaAA2A9bst5XH7exuyi5KzNMK3SEDNN8rBDnKWcUS/go-cid"
+)
+
+const (
+	colorNull trielement = iota
+	color1
+	color2
+	color3
+	colorMask = 0x3
+)
+
+const (
+	// enum from enumerator
+	// stricter enumerators should have higer value than those less strict
+	enumFast   trielement = 0
+	enumStrict trielement = 1 << 2
+	enumMask   trielement = enumStrict
+)
+
+// this const is used to enable runtime check for things that should never happen
+// will cause panic if they happen
+const pedantic = true
+
+// try to keep trielement as small as possible
+// 8bit color is overkill, 3 bits would be enough, but additional functionality
+// will probably increase it future
+// if ever it blows over 64bits total size change the colmap in triset to use
+// pointers onto this structure
+type trielement uint8
+
+func (t trielement) getColor() trielement {
+	return t & colorMask
+}
+
+func (t trielement) getEnum() trielement {
+	return t & enumMask
+}
+
+type triset struct {
+	// colors are per triset allowing fast color swap after sweep,
+	// the update operation in map of structs is about 3x as fast
+	// as insert and requres 0 allocations (keysize allocation in case of insert)
+	white, gray, black trielement
```

Comment:
how is this different from just having constants?

Reply:
https://github.com/ipfs/go-ipfs/pull/4149#discussion_r135936030